Release notes for v1.0.0-rc1 (#4635)
grafana/k6: release notes/v1.0.0-rc1.md

## UX improvements and enhancements

- [#4547](https://github.com/grafana/k6/pull/4547) The k6 banner now outputs with the original TrueColor (24-bit) logo only if the terminal supports it.
- [#4590](https://github.com/grafana/k6/pull/4590) moves a docker-compose example with InfluxDB to the `examples/docker-compose` directory and adds an opentelemetry example.
- [#4602](https://github.com/grafana/k6/pull/4602), [#4629](https://github.com/grafana/k6/pull/4629) Better error message on options error from script. k6 now will try to print the part of the options that fails as JSON.

## Bug fixes

- [#4544](https://github.com/grafana/k6/pull/4544) Fix race in ReadableStream.cancel and run WPT test with race detection for easier finding of similar problems.
- [#4567](https://github.com/grafana/k6/pull/4567), [#4558](https://github.com/grafana/k6/pull/4558), [#4574](https://github.com/grafana/k6/pull/4574), [#4610](https://github.com/grafana/k6/pull/4610) Fixes race conditions with the browser module's `ElementHandle`, `Mouse`, `Keyboard`, and `Response`.

## Maintenance and internal improvements

> - [#2770](https://github.com/grafana/k6/pull/2770) Refactors parts of the JS module.

Author: @mstoykov Is this item correct?

Reviewer: I think you did this during hte refactor of the ext.

This is the example issue in the template 
![image](https://github.com/user-attachments/assets/27c4e33c-5980-4f44-a2a3-f32e2a8045cd)
It is years old

Author: @mstoykov you're right, I didn't realize this is part of the template :facepalm: 

```suggestion
```

k6 `v1.0.0-rc1` is here 🎉! This release includes:

- A preview of the next `v1.0.0` release
- `k6/experimental/webcrypto` promoted to stable and available globally under `crypto`.

> - A revamped end-of-test summary aiming to bring an easier way to understand test results.
> 

Reviewer 2: ```suggestion
- `k6/browser` provided an API for tracking network requests and responses.

```

> - A revamped end-of-test summary aiming to bring an easier way to understand test results.
> 

Reviewer 2: ```suggestion
- `k6/browser` provided an API for tracking network requests and responses.

```

## Breaking changes

> - [#4541](https://github.com/grafana/k6/pull/4541) Commas(`,`) are now supported in the values of `--tag` cli flag. This is a breaking change as previously a comma meant the start of a new set of tag-values. As comma is a valid symbol for the value of a tag this is necessary to have equivalency between different ways of setting tags. This still allows multiple tags to be set on the cli with multiple `--tag key=value` arguments.

Author: @mstoykov It would be good to update the docs for clarifying it

Reviewer: Looking at the [doc](https://grafana.com/docs/k6/latest/using-k6/k6-options/reference/#tags) We have never said that it is possible to use `,` to give multiple values. I also looked at taht to see if we actually have ever showed that - and no.

Author: I know, and this is the point. I think we should improve it and clarify the case. We should move the various examples / documented cases to show how to use multiple tags, instead of all them on the case with only one tag.

Reviewer: https://github.com/grafana/k6-docs/pull/1909

> ### A revamped end-of-test summary aiming to bring an easier way to understand test results [#4089](https://github.com/grafana/k6/pull/4089)

Author: @joanlopez do you plan to update the documentation for it here https://grafana.com/docs/k6/latest/results-output/end-of-test?

Reviewer 3: Done here: https://github.com/grafana/k6-docs/pull/1906

INFO[0001] https://quickpizza.grafana.com/images/pizza.png  source=console
...
```

### `k6/experimental/webcrypto` promoted to stable and available globally under `crypto` [#4278](https://github.com/grafana/k6/pull/4278)

With this release, the `k6/experimental/webcrypto` module is promoted to stable and available globally under `crypto`. So you don't need to import it anymore.

```javascript
export default function () {
  const myUUID = crypto.randomUUID();

  console.log(myUUID);
}
```

`k6/experimental/webcrypto` is deprecated and will be removed in `v1.1.0`.

### Support for custom templates in `k6 new` command [#4618](https://github.com/grafana/k6/pull/4618)

The `k6 new` command now accepts a path to a file to use as a template for the new script. Templates use [Go templates syntax](https://pkg.go.dev/text/template#pkg-overview) and can include the following variables:

- `ScriptName`: The name of the new script.
- `ProjectID`: The ID of the Grafana Cloud project to use for the new script.

To generate a new script using a custom template, use the following command:

```sh
k6 new --template /path/to/my-template.js
```

> ### Secret Sources [#4514](https://github.com/grafana/k6/pull/4514), [#4621](https://github.com/grafana/k6/pull/4621)

Reviewer 4: If we have docs it's worth to put the link here

Author: https://github.com/grafana/k6-docs/pull/1905

Now k6 has a way to get secrets from different sources with a way to extend that. Among other things this means that the values received from a secret source will be redacted from the logs.

At this point the two implementations included are to read them from key-value file or from cli flags, are meant mostly for testing the feature.

It already has extension support that can be used to implement retrieving secrets from more secure sources.

> In the future k6 will likely add additional implementations that are more production ready.

Reviewer 2: It's worth adding a usage and output example here.

Author: @mstoykov what's your opinion here?

## UX improvements and enhancements

- [#4547](https://github.com/grafana/k6/pull/4547) The k6 banner now outputs with the original TrueColor (24-bit) logo only if the terminal supports it.

> - [#4590](https://github.com/grafana/k6/pull/4590) moves a docker-compose example with InfluxDB to the `examples/docker-compose` directory and adds an opentelemetry example.

Author: This is done here https://github.com/grafana/k6/pull/4635/commits/ea619ca5c7401d3f23d8dfb9044607cd419323f8

> - [#4602](https://github.com/grafana/k6/pull/4602), [#4629](https://github.com/grafana/k6/pull/4629) Better error message on options error from script. k6 now will try to print the part of the options that fails as JSON.

Reviewer 4: ```suggestion
- [#4602](https://github.com/grafana/k6/pull/4602), [#4629](https://github.com/grafana/k6/pull/4629) improves the error message on options error from script. k6 now will try to print the part of the options that fails as JSON.
```

Reviewer 2: ```suggestion
- [#4602](https://github.com/grafana/k6/pull/4602), [#4629](https://github.com/grafana/k6/pull/4629) Improves the error message on options error from script. k6 now tries to print the part of the options that fails as JSON.
```

Author: Done both

## Bug fixes

- [#4544](https://github.com/grafana/k6/pull/4544) Fix race in ReadableStream.cancel and run WPT test with race detection for easier finding of similar problems.
- [#4567](https://github.com/grafana/k6/pull/4567), [#4558](https://github.com/grafana/k6/pull/4558), [#4574](https://github.com/grafana/k6/pull/4574), [#4610](https://github.com/grafana/k6/pull/4610) Fixes race conditions with the browser module's `ElementHandle`, `Mouse`, `Keyboard`, and `Response`.

> ## Maintenance and internal improvements

Reviewer 4: How about trying to combine the items when it's possible, trying to keep essential for users on top of the list, what was discussed in our past retro.

Like, everything that related to the test or running the test in CI could go down and combined under the single list item. WDYT?

Author: Done

- [#4603](https://github.com/grafana/k6/pull/4603) Homogenize package docs comment block format.
- [#4604](https://github.com/grafana/k6/pull/4604) Bump to golangci-lint 1.64.6 - to support go 1.24.
- [#4605](https://github.com/grafana/k6/pull/4605) Refactor the code responsible for `lib.RuntimeOptions` loading.
- [#4606](https://github.com/grafana/k6/pull/4606) Bump k6 to depend on Go 1.23 and build with 1.24.
- [#4611](https://github.com/grafana/k6/pull/4611) Extract the HDR histogram implementation into a shared package.
- [#4616](https://github.com/grafana/k6/pull/4616), [#4613](https://github.com/grafana/k6/pull/4613), [#4627](https://github.com/grafana/k6/pull/4627), [#4615](https://github.com/grafana/k6/pull/4615), [#4592](https://github.com/grafana/k6/pull/4592), [#4578](https://github.com/grafana/k6/pull/4578), [#4554](https://github.com/grafana/k6/pull/4554), [#4553](https://github.com/grafana/k6/pull/4553), [#4552](https://github.com/grafana/k6/pull/4552), [#4527](https://github.com/grafana/k6/pull/4527), [#4550](https://github.com/grafana/k6/pull/4550) Update direct dependencies.

## Roadmap

> This release marks a special, non-conventional milestone in the k6 software life-cycle, serving as a preview of the upcoming version 1.0.0.

Reviewer 4: I'd suggest putting this at the top of the document. First it's align with the release highlights, secondly, but probably what's more essential, it's immediately answering the question on why after `v0.57.0` we're issuing the `v1.0.0-rc1`

Author: Done

k6 `v1.0.0-rc1` is here 🎉! This release includes:

- A preview of the next `v1.0.0` release
- `k6/experimental/webcrypto` promoted to stable and available globally under `crypto`.
- A revamped end-of-test summary aiming to bring an easier way to understand test results.
- `k6/browser` provided an API for tracking network requests and responses.

## Breaking changes

- [#4541](https://github.com/grafana/k6/pull/4541) Commas(`,`) are now supported in the values of `--tag` cli flag. This is a breaking change as previously a comma meant the start of a new set of tag-values. As comma is a valid symbol for the value of a tag this is necessary to have equivalency between different ways of setting tags. This still allows multiple tags to be set on the cli with multiple `--tag key=value` arguments.

> ### A revamped end-of-test summary aiming to bring an easier way to understand test results [#4089](https://github.com/grafana/k6/pull/4089)

Reviewer: This doesn't mention that the --summary-export is also changed by default

Author: @joanlopez can you address it, please?

Reviewer 5: Waiting on https://github.com/grafana/k6/pull/4647 to decide what to write exactly.

In the future, we'll likely include additional implementations that are more production-ready.

Here's an example where we log the secret directly, make a request, and then log the whole response. In both cases, the secrets are redacted from the logs.

```javascript
import http from 'k6/http';
import secrets from 'k6/secrets';

export default async () => {
  const my_secret = await secrets.get('cool'); // get secret from a source with the provided identifier
  console.log(my_secret);
  const response = await http.asyncRequest("GET", "https://httpbin.org/get", null, {
    headers: {
      "Custom-Authentication": `Bearer ${await secrets.get("else")}`,
    }
  })
  console.log(response.body)
}
```

```shell
$ k6 run --secret-source=mock=cool="not cool secret",else="totally a secret" script.js
...
INFO[0000] ***SECRET_REDACTED***                         source=console
INFO[0031] {
  "args": {},
  "headers": {
    "Custom-Authentication": "Bearer ***SECRET_REDACTED***",
    "Host": "httpbin.org",

>     "User-Agent": "k6/0.57.0 (https://k6.io/)",

Author: ```suggestion
    "User-Agent": "k6/1.0.0-rc1 (https://k6.io/)",
```

k6 `v1.0.0-rc1` is here 🎉!

This release marks a special, non-conventional milestone in the k6 software life-cycle, serving as a preview of the upcoming version 1.0.0.

> The purpose of this release is to give the community a chance to test the new version, identify any potential issues, and test migrations of any parts affected by breaking changes. If you encounter any problems or have trouble with the migration, we encourage you to report them by creating an issue. Your feedback will help improve the final release. If no critical issues are reported, we plan to release the final v1.0.0 version in a couple of weeks.

Reviewer 6: Just to confirm, I thought the final v1.0.0 version was going to be released in May. Is this right that it would be in a couple of weeks? 😱

Reviewer 4: @heitortsergent I believe it's a good catch :sweat_smile: The actual v1.0.0 is going to be in May, we need though to set the date in https://github.com/grafana/k6/milestone/48

Author: ```suggestion
The purpose of this release is to give the community a chance to test the new version, identify any potential issues, and test migrations of any parts affected by breaking changes. If you encounter any problems or have trouble with the migration, we encourage you to report them by creating an issue. Your feedback will help improve the final release. If no critical issues are reported, we plan to release the final `v1.0.0` within the next month.
```

Here’s a glimpse of what’s new in this release:

- `k6/experimental/webcrypto` promoted to stable and available globally under `crypto`.
- A revamped end-of-test summary aiming to bring an easier way to understand test results.

> - `k6/browser` provided an API for tracking network requests and responses.

Author: ```suggestion
- `k6/browser` provided an API for tracking network requests and responses.
- Store and retrieve secrets from different sources.
```

@mstoykov this seems important enough to be mentioned here.

> - `k6/browser` provided an API for tracking network requests and responses.

Reviewer: ```suggestion
- `k6/browser` provided an API for tracking network requests and responses.
- new `k6/secrets` module for retrieving secrets with extension support.
```

Author: ```suggestion
- `k6/browser` provided an API for tracking network requests and responses.
- The new `k6/secrets` module for retrieving secrets with extension support.
```

## Breaking changes

- [#4541](https://github.com/grafana/k6/pull/4541) Commas(`,`) are now supported in the values of the `--tag` CLI flag. This is a breaking change, as previously, a comma meant the start of a new set of tag-values. As a comma is a valid symbol for the value of a tag, this is necessary to have equivalency between different ways of setting tags. This still allows multiple tags to be set on the CLI with multiple `--tag key=value` arguments.

### A revamped end-of-test summary aiming to bring an easier way to understand test results [#4089](https://github.com/grafana/k6/pull/4089)

**TL;DR** This change is a breaking change because the data structure that [`handleSummary`](https://grafana.com/docs/k6/latest/results-output/end-of-test/custom-summary/) function receives,
has changed. So, existing custom summary handlers need to be adjusted accordingly to the new data structure. 
Meanwhile, `--summary-mode=legacy` can be used for backwards compatibility. 

Other than that, we offer no guarantees on how the summary looks like, and it might change on new releases again.

```

>  █ TOTAL RESULTS 

Reviewer 3: ```suggestion
  █ THRESHOLDS

    http_req_duration
    ✓ 'p(95)<1500' p(95)=316.31ms
    ✓ 'p(90)<2000' p(90)=297.46ms

    http_req_failed
    ✓ 'rate<0.3' rate=25.00%
 
   █ TOTAL RESULTS 
```

Reviewer 3: cc @joanlopez as I wrote the docs, I found the thresholds change is actually a big one, and it might be worth showing it here too 🙇🏻

Author: @oleiade @joanlopez this seems very big text for release notes, maybe we can move to a collapsed section? We can have a text like `Expand to see a real example of the new End-of-Test summary`. What do you think?

Reviewer 3: I've added a commit that replaces it with an image inside a collapsible section. I stored the image in a dedicated `images` folder, let me know if you'd prefer it done some other way 🙇🏻

Reviewer: I prefer this method which doesn't require us to have images in the source code that everybody pools 

https://youtu.be/2ajWUp8F694?si=uYjbe5NtgLCkwZXc&t=48

- [#4612](https://github.com/grafana/k6/pull/4612) Updates the link included in the local modules' error message. Thanks, @tanurrra!

## Bug fixes

- [#4544](https://github.com/grafana/k6/pull/4544) Fixes race in `ReadableStream.cancel` and run WPT test with race detection for easier finding of similar problems.
- [#4567](https://github.com/grafana/k6/pull/4567), [#4558](https://github.com/grafana/k6/pull/4558), [#4574](https://github.com/grafana/k6/pull/4574), [#4610](https://github.com/grafana/k6/pull/4610) Fixes race conditions with the browser module's `ElementHandle`, `Mouse`, `Keyboard`, and `Response`.
- [#4641](https://github.com/grafana/k6/pull/4641) Fixes JSON marshalling of tagset. Thank you @dvordrova 🙇.

## Maintenance and internal improvements

- [#4519](https://github.com/grafana/k6/pull/4519), [#4562](https://github.com/grafana/k6/pull/4562) Moves the Prometheus remote write, and OpenTelemetry outputs' code to the k6 repository.
- [#4545](https://github.com/grafana/k6/pull/4545) Tries to stabilize MacOS CI by making browser throttle.js easier to run for it.

> - [#4546](https://github.com/grafana/k6/pull/4546) Makes `TestConstantArrivalRateRunCorrectRate` less flaky.

Reviewer: ```suggestion
- [#4546](https://github.com/grafana/k6/pull/4546), [#4643](https://github.com/grafana/k6/pull/4643) Reduce flakiness of the test suite..
```

```javascript
import http from 'k6/http';
import secrets from 'k6/secrets';

export default async () => {
  const my_secret = await secrets.get('cool'); // get secret from a source with the provided identifier
  console.log(my_secret);
  const response = await http.asyncRequest("GET", "https://httpbin.org/get", null, {
    headers: {
      "Custom-Authentication": `Bearer ${await secrets.get("else")}`,
    }
  })
  console.log(response.body)
}
```

```shell
$ k6 run --secret-source=mock=cool="not cool secret",else="totally a secret" script.js
...
INFO[0000] ***SECRET_REDACTED***                         source=console
INFO[0031] {
  "args": {},
  "headers": {
    "Custom-Authentication": "Bearer ***SECRET_REDACTED***",
    "Host": "httpbin.org",
    "User-Agent": "k6/0.57.0 (https://k6.io/)",
    "X-Amzn-Trace-Id": "Root=1-67dd6691-18eeaf5d1782bf292da5037c"
  },
  "origin": "1.1.1.1",
  "url": "https://httpbin.org/get"
}  source=console
...
```

## UX improvements and enhancements

- [#4547](https://github.com/grafana/k6/pull/4547) The k6 banner now outputs with the original TrueColor (24-bit) logo only if the terminal supports it.

> - [#4590](https://github.com/grafana/k6/pull/4590) Moves a docker-compose example with InfluxDB to the `examples/docker-compose` directory and adds an opentelemetry example.

Author: ```suggestion
- [#4590](https://github.com/grafana/k6/pull/4590) Moves the `docker-compose` example with InfluxDB to the `examples/docker-compose` directory and adds an opentelemetry example.
```

- [#4602](https://github.com/grafana/k6/pull/4602), [#4629](https://github.com/grafana/k6/pull/4629) Improves the error message on options error from script. k6 now will try to print the part of the options that fails as JSON.
- [#4612](https://github.com/grafana/k6/pull/4612) Updates the link included in the local modules' error message. Thanks, @tanurrra!

## Bug fixes

> - [#4544](https://github.com/grafana/k6/pull/4544) Fix race in `ReadableStream.cancel` and run WPT test with race detection for easier finding of similar problems.

Reviewer 4: ```suggestion
- [#4544](https://github.com/grafana/k6/pull/4544) Fixes race in `ReadableStream.cancel` and run WPT test with race detection for easier finding of similar problems.
```

- [#4567](https://github.com/grafana/k6/pull/4567), [#4558](https://github.com/grafana/k6/pull/4558), [#4574](https://github.com/grafana/k6/pull/4574), [#4610](https://github.com/grafana/k6/pull/4610) Fixes race conditions with the browser module's `ElementHandle`, `Mouse`, `Keyboard`, and `Response`.

> - [#4641](https://github.com/grafana/k6/pull/4641) Fix JSON marshalling of tagset. Thank you @dvordrova 🙇.

Reviewer 4: ```suggestion
- [#4641](https://github.com/grafana/k6/pull/4641) Fixes JSON marshalling of tagset. Thank you @dvordrova 🙇.
```

## Maintenance and internal improvements

> - [#4519](https://github.com/grafana/k6/pull/4519), [#4562](https://github.com/grafana/k6/pull/4562) Moves the Prometheus remote write, and OpenTelemetry outputs' code to the k6 repository.
> - [#4545](https://github.com/grafana/k6/pull/4545) Trying to stabilize MacOS CI by making browser throttle.js easier to run for it.
> - [#4546](https://github.com/grafana/k6/pull/4546) Make `TestConstantArrivalRateRunCorrectRate` less flaky.
> - [#4552](https://github.com/grafana/k6/pull/4552) Updates the fallback x509 certificates for CA roots.
> - [#4561](https://github.com/grafana/k6/pull/4561), [#4638](https://github.com/grafana/k6/pull/4638) Update Sobek to fix a bug in `Function.apply`, logical assignment support, array destructuring in exports, better cross-os source map support.
> - [#4563](https://github.com/grafana/k6/pull/4563) Fix `k6/timers` reporting as always being used.
> - [#4603](https://github.com/grafana/k6/pull/4603) Homogenize package docs comment block format.
> - [#4605](https://github.com/grafana/k6/pull/4605) Refactor the code responsible for `lib.RuntimeOptions` loading.
> - [#4606](https://github.com/grafana/k6/pull/4606), [#4604](https://github.com/grafana/k6/pull/4604) Bump k6 to depend on greater or equal Go 1.23 versions, build with 1.24, then use a compatible golangci-lint version. 
> - [#4611](https://github.com/grafana/k6/pull/4611) Extract the HDR histogram implementation into a shared package.
> - [#4616](https://github.com/grafana/k6/pull/4616), [#4613](https://github.com/grafana/k6/pull/4613), [#4627](https://github.com/grafana/k6/pull/4627), [#4615](https://github.com/grafana/k6/pull/4615), [#4592](https://github.com/grafana/k6/pull/4592), [#4578](https://github.com/grafana/k6/pull/4578), [#4554](https://github.com/grafana/k6/pull/4554), [#4553](https://github.com/grafana/k6/pull/4553), [#4552](https://github.com/grafana/k6/pull/4552), [#4527](https://github.com/grafana/k6/pull/4527), [#4550](https://github.com/grafana/k6/pull/4550) Update direct dependencies.

Reviewer 4: Diff is not perfect, but I'm suggesting to combine two test related PR and put them down of the list since it's not so relevant for release notes readers 

```suggestion
- [#4519](https://github.com/grafana/k6/pull/4519), [#4562](https://github.com/grafana/k6/pull/4562) Moves the Prometheus remote write, and OpenTelemetry outputs' code to the k6 repository.
- [#4552](https://github.com/grafana/k6/pull/4552) Updates the fallback x509 certificates for CA roots.
- [#4561](https://github.com/grafana/k6/pull/4561), [#4638](https://github.com/grafana/k6/pull/4638) Update Sobek to fix a bug in `Function.apply`, logical assignment support, array destructuring in exports, better cross-os source map support.
- [#4563](https://github.com/grafana/k6/pull/4563) Fix `k6/timers` reporting as always being used.
- [#4603](https://github.com/grafana/k6/pull/4603) Homogenize package docs comment block format.
- [#4605](https://github.com/grafana/k6/pull/4605) Refactor the code responsible for `lib.RuntimeOptions` loading.
- [#4606](https://github.com/grafana/k6/pull/4606), [#4604](https://github.com/grafana/k6/pull/4604) Bump k6 to depend on greater or equal Go 1.23 versions, build with 1.24, then use a compatible golangci-lint version. 
- [#4611](https://github.com/grafana/k6/pull/4611) Extract the HDR histogram implementation into a shared package.
- [#4616](https://github.com/grafana/k6/pull/4616), [#4613](https://github.com/grafana/k6/pull/4613), [#4627](https://github.com/grafana/k6/pull/4627), [#4615](https://github.com/grafana/k6/pull/4615), [#4592](https://github.com/grafana/k6/pull/4592), [#4578](https://github.com/grafana/k6/pull/4578), [#4554](https://github.com/grafana/k6/pull/4554), [#4553](https://github.com/grafana/k6/pull/4553), [#4552](https://github.com/grafana/k6/pull/4552), [#4527](https://github.com/grafana/k6/pull/4527), [#4550](https://github.com/grafana/k6/pull/4550) Update direct dependencies.
- [#4545](https://github.com/grafana/k6/pull/4545), [#4546](https://github.com/grafana/k6/pull/4546) improve test.
```

Reviewer 2: 👍

Some (Fix, Refactor, etc.) don't use the present tense like others (Moves, Updates, etc.).

Reviewer 4: >Some (Fix, Refactor, etc.) don't use the present tense like others (Moves, Updates, etc.).

yeah, I do agree, but it's worth probably to align them everywhere at once, @codebien could you please take care of it?

> - [#4639](https://github.com/grafana/k6/pull/4639) Pushes the `latest` docker image tag even for `rc` releases.

Author: @mstoykov we would prefer to group similar items in maintenance section. This change seems to be grouped with `4546 Make TestConstantArrivalRateRunCorrectRate less flaky` in something like `Reduce flakiness of the test suite`.

> - [#4639](https://github.com/grafana/k6/pull/4639) Pushes the `latest` docker image tag even for `rc` releases.

Reviewer: I also wanted that , but htere were already changes to this line from other comments and if I added mine it will not work. 

Maybe let's merge the others and we can add this on top of them after, and keep this as a reminder until then

> - [#4639](https://github.com/grafana/k6/pull/4639) Pushes the `latest` docker image tag even for `rc` releases.

Reviewer: https://github.com/grafana/k6/pull/4635/files#r2010022644

### A new default path for the configuration file [#4301](https://github.com/grafana/k6/pull/4301)

When running the `k6 cloud login` or the deprecated `k6 login` commands, a configuration file was automatically created at `{USER_CONFIG_DIR}/loadimpact/config.json`. Now, the configuration file is created at `{USER_CONFIG_DIR}/k6/config.json`.

To migrate your configuration file to the new path:

1. Run `k6 cloud login` or `k6 login` to automatically migrate the configuration file to the new location.
2. Run `k6 cloud run` or `k6 run` to verify that the version is now fully functional and no related warning is emitted.

The configuration file in the old path remains available and can continue to be used with the previous k6 versions. If you're not using an old version of k6 anymore, consider deleting the files manually.

The `k6 run` commands search for the configuration file in the new location. If it can't find it, it tries to fall back on the old path and then logs a warning message suggesting to migrate it.

## New features

> ### A revamped end-of-test summary aiming to bring an easier way to understand test results [#4089](https://github.com/grafana/k6/pull/4089), [#4649](https://github.com/grafana/k6/pull/4649)

Reviewer 5: @heitortsergent finally we moved this section within `New features`, because a recent change -[#4649](https://github.com/grafana/k6/pull/4649)- made it no longer being a breaking change.

I applied most of your suggestions, but of course a few no longer apply. Note that I used the style you suggested for the "note", but now I moved it to the end of the section, as it is way less relevant now.

Reviewer 6: Sounds good, thanks for letting me know. 🙇

```sh
k6 new --template /path/to/my-template.js
```

> ### [Secret Sources](https://grafana.com/docs/k6/latest/using-k6/secret-source/) [#4514](https://github.com/grafana/k6/pull/4514), [#4621](https://github.com/grafana/k6/pull/4621), [#4637](https://github.com/grafana/k6/pull/4637)

Author: ```suggestion
### Secret Sources [#4514](https://github.com/grafana/k6/pull/4514), [#4621](https://github.com/grafana/k6/pull/4621), [#4637](https://github.com/grafana/k6/pull/4637)
```
@mstoykov If you check the rendered version, it's a bit strange because it seems a unique link, where instead we have a few different. I prefer the classic form of mentioning the documentation later.

> Now k6 has a way to get secrets from different sources with a way to extend that. Among other things this means that the values received from a secret source will be redacted from the logs.

Author: ```suggestion
Now k6 has a way to get secrets from different sources with a way to extend that. Among other things this means that the values received from a secret source will be redacted from the logs. See the [documentation](https://grafana.com/docs/k6/latest/using-k6/secret-source) for more information.
```
